Remotion/ObjectBinding/Sample/SessionStateReflectionBusinessObjectStorageProvider.cs

```diff
@@ -0,0 +1,86 @@
+// 
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Remotion.Utilities;
+using Remotion.Web.Infrastructure;
+
+namespace Remotion.ObjectBinding.Sample
+{
+  public class SessionStateReflectionBusinessObjectStorageProvider : IReflectionBusinessObjectStorageProvider
+  {
+    private static readonly string s_sessionStateEntryName = typeof (SessionStateReflectionBusinessObjectStorageProvider).FullName;
+
+    private readonly IHttpContextProvider _httpContextProvider;
+    private readonly IReflectionBusinessObjectStorageProviderFactory _reflectionBusinessObjectStorageProviderFactory;
+
+    public SessionStateReflectionBusinessObjectStorageProvider (
+        IHttpContextProvider httpContextProvider,
+        IReflectionBusinessObjectStorageProviderFactory reflectionBusinessObjectStorageProviderFactory)
+    {
+      ArgumentUtility.CheckNotNull ("httpContextProvider", httpContextProvider);
+      ArgumentUtility.CheckNotNull ("reflectionBusinessObjectStorageProviderFactory", reflectionBusinessObjectStorageProviderFactory);
+
+      _httpContextProvider = httpContextProvider;
+      _reflectionBusinessObjectStorageProviderFactory = reflectionBusinessObjectStorageProviderFactory;
+    }
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<Guid> GetObjectIDsForType (Type type)
+    {
+      ArgumentUtility.CheckNotNull ("type", type);
+
+      var reflectionBusinessObjectStorageProvider = GetReflectionBusinessObjectStorageProviderForCurrentSession();
+      return reflectionBusinessObjectStorageProvider.GetObjectIDsForType (type);
+    }
+
+    /// <inheritdoc />
+    public Stream GetReadObjectStream (Type type, Guid id)
+    {
+      ArgumentUtility.CheckNotNull ("type", type);
+
+      var reflectionBusinessObjectStorageProvider = GetReflectionBusinessObjectStorageProviderForCurrentSession();
+      return reflectionBusinessObjectStorageProvider.GetReadObjectStream (type, id);
+    }
+
+    /// <inheritdoc />
+    public Stream GetWriteObjectStream (Type type, Guid id)
+    {
+      ArgumentUtility.CheckNotNull ("type", type);
+
+      var reflectionBusinessObjectStorageProvider = GetReflectionBusinessObjectStorageProviderForCurrentSession();
+      return reflectionBusinessObjectStorageProvider.GetWriteObjectStream (type, id);
+    }
+
+    private IReflectionBusinessObjectStorageProvider GetReflectionBusinessObjectStorageProviderForCurrentSession ()
+    {
+      var httpContext = _httpContextProvider.GetCurrentHttpContext();
+      var httpSessionStateBase = httpContext.Session;
+
+      var reflectionBusinessObjectStorageProvider = httpSessionStateBase?[s_sessionStateEntryName] as IReflectionBusinessObjectStorageProvider;
+      if (reflectionBusinessObjectStorageProvider == null)
+      {
+        reflectionBusinessObjectStorageProvider = _reflectionBusinessObjectStorageProviderFactory.CreateBusinessObjectStorageProvider();
+        if (httpSessionStateBase != null)
```

Comment:
could we write this as `httpSession?[s_sessionKeyForStorageProvider] = reflectionBusinessObjectStorageProvider;` ? I.e. using null-propagation also for the write side.

Reply:
I have tried but this is not possible :/